File: algorithms/java/src/test/java/leetcode/LC081SearchInRotatedSortedArrayII.java
package leetcode;

import org.junit.jupiter.api.Test;


import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/search-in-rotated-sorted-array-ii/
 * #medium
 */
public final class LC081SearchInRotatedSortedArrayII {
    public final class Solution {
        public boolean search(final int[] nums, final int target) {
            for (int i = 1; i < nums.length; i++) {
                if (nums[i - 1] > nums[i]) {
                    boolean ante = Arrays.binarySearch(nums, 0, i, target) >= 0;
                    boolean post = Arrays.binarySearch(nums, i, nums.length, target) >= 0;
                    return ante || post;
                }
            }
            return Arrays.binarySearch(nums, 0, nums.length, target) >= 0;
        }
    }

    @Test
    public void test_simple_example() {
        int[] nums = {4, 5, 6, 6, 7, 0, 1, 2};
        assertEquals(true, new Solution().search(nums, 1));
    }

    @Test
    public void test_tricky_example() {
        int[] nums = {1, 1, 3, 1, 1, 1, 1, 1};
        assertEquals(false, new Solution().search(nums, 2));
    }
}

File: algorithms/code/leetcode/lc081_search_in_rotated_sorted_array_ii/lc081_search_in_rotated_sorted_array_ii.py
# https://leetcode.com/problems/search-in-rotated-sorted-array-ii/

import bisect
import unittest


class Solution:
    @staticmethod
    def __binary_search(
        array: list[int], begin: int, end: int, value: int
    ) -> int:
        index = bisect.bisect_left(array, value, begin, end)
        if index != len(array) and array[index] == value:
            return index
        return -1

    def search(self, nums: list[int], target: int) -> bool:
        for i in range(1, len(nums)):
            if nums[i - 1] > nums[i]:
                ante = self.__binary_search(nums, 0, i, target) >= 0
                post = self.__binary_search(nums, i, len(nums), target) >= 0
                return ante or post
        return self.__binary_search(nums, 0, len(nums), target) >= 0


class TestCode(unittest.TestCase):
    def test_simple_example(self) -> None:
        nums = [4, 5, 6, 6, 7, 0, 1, 2]
        self.assertTrue(Solution().search(nums, 1))

    def test_tricky_example(self) -> None:
        nums = [1, 1, 3, 1, 1, 1, 1, 1]
        self.assertFalse(Solution().search(nums, 2))

    def test_1_and_1(self) -> None:
        nums = [1]
        self.assertTrue(Solution().search(nums, 1))

    def test_1_and_0(self) -> None:
        nums = [1]
        self.assertFalse(Solution().search(nums, 0))

    def test_2_5_6_0_0_1_2__3(self) -> None:
        nums = [2, 5, 6, 0, 0, 1, 2]
        self.assertFalse(Solution().search(nums, 3))

    def test_2_5_6_0_0_1_2__0(self) -> None:
        nums = [2, 5, 6, 0, 0, 1, 2]
        self.assertTrue(Solution().search(nums, 0))

    def test_2_2_2_3_2_2_2__3(self) -> None:
        nums = [2, 2, 2, 3, 2, 2, 2]
        self.assertTrue(Solution().search(nums, 3))
